File: exp1.py

class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        self.height = 1

def height(node):
    if not node:
        return 0
    return node.height

def get_balance(node):
    if not node:
       return 0
    return height(node.left) - height(node.right)

def update_height(node):
    if node:
       node.height = 1 + max(height(node.left), height(node.right))

def rotate_right(y):
    x = y.left
    T2 = x.right
    x.right = y
    y.left = T2
    update_height(y)
    update_height(x)
    return x

def rotate_left(x):
    y = x.right
    T2 = y.left
    y.left = x
    x.right = T2
    update_height(x)
    update_height(y)
    return y

def insert(node, key):
    if not node:
       return Node(key)
    if key < node.key:
        node.left = insert(node.left, key)
    elif key > node.key:
        node.right = insert(node.right, key)
    else:
        return node

    update_height(node)
    balance = get_balance(node)

    if balance > 1:
        if key < node.left.key:
            return rotate_right(node)
        else:
            node.left = rotate_left(node.left)
            return rotate_right(node)
    if balance < -1:
        if key > node.right.key:
            return rotate_left(node)
        else:
            node.right = rotate_right(node.right)
            return rotate_left(node)
    return node

def min_value_node(node):
    current = node
    while current.left:
        current = current.left
    return current

def delete_node(root, key):
    if not root:
        return root
    if key < root.key:
        root.left = delete_node(root.left, key)
    elif key > root.key:
        root.right = delete_node(root.right, key)
    else:
        if not root.left:
            return root.right
        elif not root.right:
            return root.left
        temp = min_value_node(root.right)
        root.key = temp.key
        root.right = delete_node(root.right, temp.key)

    update_height(root)
    balance = get_balance(root)

    if balance > 1:
        if get_balance(root.left) >= 0:
            return rotate_right(root)
        else:
            root.left = rotate_left(root.left)
            return rotate_right(root)
    if balance < -1:
        if get_balance(root.right) <= 0:
            return rotate_left(root)
        else:
            root.right = rotate_right(root.right)
            return rotate_left(root)
    return root

def inorder_traversal(node):
    if not node:
        return []
    return inorder_traversal(node.left) + [node.key] + inorder_traversal(node.right)

def process_file(filename):
    root = None
    with open(filename, 'r') as file:
        lines = file.readlines()
        if len(lines) < 2:
            print("File should contain at least two lines.")
            return None

        insert_values = [int(x.strip()) for x in lines[0].split(',')]
        for value in insert_values:
            root = insert(root, value)

        print("After insertion:", inorder_traversal(root))

        delete_values = [int(x.strip()) for x in lines[1].split(',')]
        for value in delete_values:
            root = delete_node(root, value)

        print("After deletion:", inorder_traversal(root))

    return root

def write_avl_to_file(root, filename):
    with open(filename, 'w') as file:
        inorder_to_file(root, file)

def inorder_to_file(root, file, is_first=[True]):
    if root:
        inorder_to_file(root.left, file, is_first)
        if not is_first[0]:
            file.write(",")
        else:
            is_first[0] = False
        file.write(str(root.key))
        inorder_to_file(root.right, file, is_first)

input_filename = "C:\ADS&AA LAB\input.txt"
output_filename = "C:\ADS&AA LAB\output.txt"

root = process_file(input_filename)
if root:
    write_avl_to_file(root, output_filename)
    print(f"Final AVL tree has been written to '{output_filename}'")

File: exp1.c
#include <stdio.h>
#include <stdlib.h>

// ------------------------ AVL Tree Node Definition ------------------------

typedef struct Node {
    int key;
    struct Node *left;
    struct Node *right;
    int height;
} Node;

// ------------------------ Utility Functions ------------------------

int height(Node *N) {
    return (N == NULL) ? 0 : N->height;
}

int max(int a, int b) {
    return (a > b) ? a : b;
}

Node* newNode(int key) {
    Node* node = (Node*) malloc(sizeof(Node));
    node->key = key;
    node->left = node->right = NULL;
    node->height = 1;
    return node;
}

// ------------------------ Rotations ------------------------

Node* rightRotate(Node *y) {
    Node *x = y->left;
    Node *T2 = x->right;
    x->right = y;
    y->left = T2;
    y->height = max(height(y->left), height(y->right)) + 1;
    x->height = max(height(x->left), height(x->right)) + 1;
    return x;
}

Node* leftRotate(Node *x) {
    Node *y = x->right;
    Node *T2 = y->left;
    y->left = x;
    x->right = T2;
    x->height = max(height(x->left), height(x->right)) + 1;
    y->height = max(height(y->left), height(y->right)) + 1;
    return y;
}

// ------------------------ Balancing ------------------------

int getBalance(Node *N) {
    return (N == NULL) ? 0 : height(N->left) - height(N->right);
}

// ------------------------ Insert Function ------------------------

Node* insert(Node* node, int key) {
    if (node == NULL) return newNode(key);

    if (key < node->key)
        node->left = insert(node->left, key);
    else if (key > node->key)
        node->right = insert(node->right, key);
    else
        return node;

    node->height = 1 + max(height(node->left), height(node->right));

    int balance = getBalance(node);

    if (balance > 1 && key < node->left->key)
        return rightRotate(node);

    if (balance < -1 && key > node->right->key)
        return leftRotate(node);

    if (balance > 1 && key > node->left->key) {
        node->left = leftRotate(node->left);
        return rightRotate(node);
    }

    if (balance < -1 && key < node->right->key) {
        node->right = rightRotate(node->right);
        return leftRotate(node);
    }

    return node;
}

// ------------------------ Delete Function ------------------------

Node *minValueNode(Node* node) {
    Node* current = node;
    while (current->left != NULL)
        current = current->left;
    return current;
}

Node* deleteNode(Node* root, int key) {
    if (root == NULL) return root;

    if (key < root->key)
        root->left = deleteNode(root->left, key);
    else if (key > root->key)
        root->right = deleteNode(root->right, key);
    else {
        if ((root->left == NULL) || (root->right == NULL)) {
            Node *temp = root->left ? root->left : root->right;
            if (temp == NULL) {
                temp = root;
                root = NULL;
            } else *root = *temp;
            free(temp);
        } else {
            Node* temp = minValueNode(root->right);
            root->key = temp->key;
            root->right = deleteNode(root->right, temp->key);
        }
    }

    if (root == NULL) return root;

    root->height = 1 + max(height(root->left), height(root->right));

    int balance = getBalance(root);

    if (balance > 1 && getBalance(root->left) >= 0)
        return rightRotate(root);

    if (balance > 1 && getBalance(root->left) < 0) {
        root->left = leftRotate(root->left);
        return rightRotate(root);
    }

    if (balance < -1 && getBalance(root->right) <= 0)
        return leftRotate(root);

    if (balance < -1 && getBalance(root->right) > 0) {
        root->right = rightRotate(root->right);
        return leftRotate(root);
    }

    return root;
}

// ------------------------ In-order Traversal to File ------------------------

void inorderToFile(Node* root, FILE *fp) {
    if (root != NULL) {
        inorderToFile(root->left, fp);
        fprintf(fp, "%d ", root->key);
        inorderToFile(root->right, fp);
    }
}

// ------------------------ Main Function (Console Application Entry Point) ------------------------

int main() {
    Node* root = NULL;
    FILE *infile = fopen("C:\ads aa c\Input.txt", "r");
    FILE *outfile;
    int num;

    if (infile == NULL) {
        printf("Input file not found.\n");
        return 1;
    }

    while (fscanf(infile, "%d", &num) != EOF) {
        root = insert(root, num);
    }
    fclose(infile);

    // Sample insert and delete operations
    root = insert(root, 15);
    root = deleteNode(root, 10);

    outfile = fopen("output.txt", "w");
    inorderToFile(root, outfile);
    fclose(outfile);
    printf("AVL tree constructed and written to output.txt\n");
    return 0;
}


